### PR TITLE
GH1739: NuGetContentResolver finds assemblies in root

### LINKED ISSUE #1739: NuGetContentResolver can't find assemblies if located in root
This is indeed an error with the package. But old V2 Content resolver could resolve packages without and TFM.

Issue here:
https://github.com/cake-build/cake/blob/develop/src/Cake.NuGet/NuGetContentResolver.cs#L106

Change that to a warning and just return assembly from root.

### PR DESCRIPTION
**Pull request by mholo65**
- Will treat root assemblies as AnyFramework
- Fixes #1739

### CODE REVIEW

**Comment by mholo65** [comment]
Will add some tests tomorrow

File: src/Cake.NuGet/NuGetContentResolver.cs [hunk 1]
```diff
@@ -72,6 +76,11 @@ private IReadOnlyCollection<IFile> GetAddinAssemblies(DirectoryPath path, Packag
                 // Get relative path.
                 var relative = path.GetRelativePath(assembly.Path);
                 var framework = ParseFromDirectoryPath(current, relative.GetDirectory());
+                if (framework == null)
+                {
+                    _log.Warning("Unable to parse compatible framework from path {0}. Setting any framework.", assembly.Path.MakeAbsolute(_environment).FullPath);
```
Comment (mholo65) [inline src/Cake.NuGet/NuGetContentResolver.cs]: I'd like to hear @gep13's opinion on the Warning message.

Comment (gep13) [inline src/Cake.NuGet/NuGetContentResolver.cs]: @mholo65 only slight comment would be the last part `Setting any framework`.  I know what you are trying to say, but not sure if that works.  Perhaps something like `, as a result, no framework will be specified.`

Comment (patriksvensson) [inline src/Cake.NuGet/NuGetContentResolver.cs]: @mholo65 I'm not @gep13 (obviously), but I think that `Setting any framework.` sounds a little bit "leaky" in the sense that we leak internal information to the consumer which probably won't understand what we're talking about.

Perhaps we could change it to something similar to `Falling back to using root folder of NuGet package.` or something like that.

Thoughts @gep13?

Comment (mholo65) [inline src/Cake.NuGet/NuGetContentResolver.cs]: In this particular context, saying `falling back to using root folder of NuGet package.` is not the truth. When we can't parse framework from the path (be it `foo/bar/lib.dll` or just `lib.dll` we will treat it as compatible by setting framework to `AnyFramework`. By setting that, we can try loading it if we can't find a better match (e.g. `lib/net45/...`, `lib/net46/...`)

Comment (patriksvensson) [inline src/Cake.NuGet/NuGetContentResolver.cs]: @mholo65 What will the root folder for searching be with `AnyFramework`?

Comment (mholo65) [inline src/Cake.NuGet/NuGetContentResolver.cs]: It's actually the other way round. For each assembly, we will take the directory and try to parse the TFM and map that with the assemblies in that directory. After looping through all assemblies, we will find the closest matching our Target Framework (.NET Framework, Version=4.6 or .NET Standard, Version=1.6) and return all files mapped to that.

Now when looking over code once more, I found a smarter way to log warning. Now we log for each assembly found in root. Instead, we should only log if assemblies mapped to `AnyFramework` is actually returned. Let me modify this PR slightly :smile:

**Comment by mholo65** [comment]
image

Example of Warning message. It's _almost_ the truth. We are now treating root folder as **AnyFramework**, but fact is that if path to assembly contains **any** (e.g. `lib/any/addin.dll`) it would also be parsed as **AnyFramework**. So *Falling back to using root folder...* might not be 100% accurate.

See **NuGetFramework.Parse(string)** Tests [here](https://github.com/NuGet/NuGet.Client/blob/4cccb13833ad29d6a0bcff055460d964f1b49cfe/test/NuGet.Core.Tests/NuGet.Frameworks.Test/SpecialFrameworkTests.cs#L12) 

Thought? @patriksvensson @gep13

**Comment by patriksvensson** [comment]
@mholo65 I think it's close enough :) Haven't even seen the `any`, `agnostic` or `unsupported` convention before.

**Comment by devlead** [comment]
@mholo65  your changes have been merged, thanks for your contribution 👍

**Pull request merged**